Contract monitor, test generation pipeline

Jabe03 opened the pull request:
### Added:
- New mode: Contract Monitor
  - Same as interpreter, except the (internal) equations associated to the contract are left, the top node is abstracted, and the outputs are expected to be given in the input file
  - Output is also similar to interpreter, except now shows a detailed trace of the contract which shows the truth values of modes and guarantees at each instant  
 
###  Changed: 
- Interpreter
    - In XML and JSON, additionally outputs detailed information about the contract (like contract monitor)
    - Removed old section in JSON and XML for active modes
      - If desired by a user, the same information can be extracted from the required mode's truth values
    - Pretty-print to console unchanged
- Testgen
  - Now outputs JSON in the same format the interpreter uses for input
    - Designed to be easily consumed by some external test execution engine, or by Kind 2 itself
  - Deleted code within test generation that generated the oracle
- Pretty-printing boolean values
  - "true" is now "tt" everywhere it is printed as part of an execution path when printing to console ("false" -> "ff")

Jabe03 committed:
Added data support for contract monitor, mirrors interpreter functionality currently, input trace not functional

Jabe03 committed:
Formatting correction

Jabe03 committed:
Testgen outputs json (this breaks the oracle)

Jabe03 committed:
Contract monitor implemented (TODO: option for full contract for json and xml in lustrePath)

Jabe03 committed:
in pretty printing to the console, "true" is now "tt" and "false" is now "ff"

Jabe03 committed:
Testgen outputs JSON that is a valid input file for the interpreter

Jabe03 committed:
Cleaned up, commented, merged common functionality in lustrePath

Jabe03 committed:
Merge branch 'kind2-mc:develop' into josh-contract-monitor

Jabe03 committed:
contract-monitor ==> contract_monitor

Jabe03 committed:
Removed xml and json active-modes sections

Jabe03 committed:
Removed unused vars

Jabe03 committed:
More unused var removals

Jabe03 committed:
More cleanup

daniel-larraz commented on src/flags.mli:
```diff
@@ -775,6 +775,17 @@ module Interpreter : sig
+
+  (* (** Run number of steps, override the number of steps given in the input
+    file. *)
+  val steps : unit -> int *)
```
Perhaps we could offer this option to the contract monitor as well. It should be straightforward, right?

daniel-larraz commented on src/flags.ml:
```diff
@@ -2610,6 +2610,53 @@ module Interpreter = struct
+    Format.fprintf fmt "@[<v>\
+      The contract monitor is a special mode where Kind 2 reads a @ \ 
+model trace (inputs and outputs) from a file and prints the @\ 
+truth values of the modes and guaruntees of the system at each step.\
```
Typo: `guaruntees`

daniel-larraz commented on src/analysis.ml:
```diff
@@ -413,6 +419,7 @@ let pp_print_param_of_result pp_print_system_user_name fmt { param ; sys } =
   let param = shrink_param_to_sys param sys in
   match param with
   | Interpreter _ -> Format.fprintf fmt "simulating system"
+  | ContractMonitor _ -> Format.fprintf fmt "simulating system"
```
Replace `"simulating system"` with `"monitoring contract"`.

daniel-larraz commented on src/analysis.mli:
```diff
@@ -98,6 +98,10 @@ type param =
   (** Refinement of a system. Store the result of the previous analysis. *)
+  
+  | ContractMonitor of info 
+  (* Monitoring a trace of execution, leaving the contract as proof obligations *)
```
I would replace this description with:
> Monitoring the satisfaction of a contract for an input trace.

daniel-larraz commented on src/analysis.ml:
```diff
@@ -93,6 +93,8 @@ type param =
   (* Refinement of a system. Store the result of the previous analysis. *)
   | Refinement of info * result
+  (* Monitoring a trace of execution, leaving the contract as proof obligations *)
+  | ContractMonitor of info 
```
I would replace this description with:
> Monitoring the satisfaction of a contract for an input trace.

daniel-larraz commented on src/inputParser.ml:
```diff
@@ -180,7 +180,7 @@ let record_to_tuple assoc =
    and return the associated assignments.
    It can return multiple variable assignements if the value is an array/record/tuple. *)
-let rec read_val scope name indexes arr_indexes json =
+let rec read_val ?(only_inputs = true) scope name indexes arr_indexes json  = (* Need to somehow make this able to accept outputs as well*)
```
Remove the personal comment at the end of the line.

daniel-larraz commented on src/kEvent.ml:
```diff
@@ -500,23 +500,21 @@ let pp_print_trace_pt ?(title="Counterexample") ?(color="red")
   Format.fprintf ppf 
     "%a"
-    (InputSystem.pp_print_path_pt input_sys trans_sys true)
+    (InputSystem.pp_print_path_pt ~full_contract:full_contract input_sys trans_sys true)
```
Tip: If the name of the argument is the same as the name of the parameter, you can just write `~full_contract`.

Jabe03 replied:
Thank you!

daniel-larraz commented on src/inputSystem.mli:
```diff
@@ -84,14 +86,17 @@ val trans_sys_of_analysis:
 (** Output a path in the input system *)
 val pp_print_path_json : _ t -> TransSys.t -> bool -> Format.formatter -> Model.path -> unit
 
+(** Output a path in the input system *)
```
Write a brief comment explaining how this function differs from `pp_print_path_json`.

daniel-larraz commented on src/contractMonitor.ml:
```diff
@@ -0,0 +1,273 @@
+(* This file is part of the Kind 2 model checker.
```
Since the contract monitor shares most of its code with the interpreter, I'd rather introduce a flag in the interpreter to run it as the contract monitor than create a separate module for it.

daniel-larraz commented on src/testgen/testgenIO.ml:
```diff
@@ -155,27 +155,32 @@ let testcase_count (type s) : s t -> int
     (InputSystem.pp_print_path_in_csv in_sys sys true)
     (Model.path_from_model (TransSys.state_vars sys) cex k)
 
+(* Converts a model to the system's input values in csv. *)  
```
```suggestion
(* Converts a model to the system's input values in json. *)  
```

daniel-larraz commented on src/lustre/lustrePath.ml:
```diff
@@ -37,7 +37,15 @@ type t =
+                                 its purpose is to keep track of which mode was active 
+                                 at each timestep, but this information is captured in 
+                                 the truth values of the "required modes" part of this 
+                                 datatype*)
```
Is there any pending task that requires action? Please update the comment accordingly.

daniel-larraz commented on src/lustre/lustrePath.ml:
```diff
@@ -451,13 +458,185 @@ let active_modes_to_strings =
   in
   loop [] [] []
 
+let guaruntees_of_instances model_top instances = function
```
Typo: `guaruntees`

daniel-larraz commented on src/lustre/lustrePath.ml:
```diff
@@ -1802,21 +2046,33 @@ let rec pp_print_lustre_path_json' is_top const_map ppf = function
+      let rec interleave = function 
+      | (l, []) | ([], l) -> l 
+      | (x::xs, y::ys) -> x::y::(interleave (xs,ys))
+      in
```
This function is duplicated; it is also defined globally. Remove this instance and use the global definition.

daniel-larraz commented on src/lustre/lustrePath.ml:
```diff
@@ -955,9 +1141,18 @@ let pp_print_stream_section_pt ident_width val_width sect ppf = function
+    "%a%a%a"
+    (pp_print_stream_section_pt ident_width val_width "Assumptions") (List.map (function (name, stream_type, mode_trace) -> (name, stream_type, List.map (function (tr_val) -> (Some tr_val)) mode_trace) ) assumed)
+    (pp_print_stream_section_pt ident_width val_width "Guarantees") (List.map (function (name, stream_type, mode_trace) -> (name, stream_type, List.map (function (tr_val) -> (Some tr_val)) mode_trace) ) guarunteed)
+    (pp_print_stream_section_pt ident_width val_width "Modes") (List.map (function (name, stream_type, mode_trace) -> (name, stream_type, List.map (function (tr_val) -> (Some tr_val)) mode_trace) ) modes)
```
These lines are very long. Define the lists computed with `map` separately and use them here.

daniel-larraz commented on src/lustre/lustrePath.ml:
```diff
@@ -1850,6 +2106,47 @@ let pp_print_path_json
+    pp_print_lustre_path_json' false const_map ppf tl
+
+
+  let pp_print_lustre_path_json_testgen ppf (path, const_map) =
```
This line should not be indented.

daniel-larraz commented on src/lustre/lustrePath.ml:
```diff
@@ -1850,6 +2106,47 @@ let pp_print_path_json
+  Format.fprintf ppf "[@[<v 1>%a@]]"
+    (pp_print_lustre_path_json_testgen' const_map) [path]
+
+  let pp_print_path_json_testgen
```
This line should not be indented.

daniel-larraz reviewed (changes requested):
Please address the comments I left. Additionally, update the Test Generation section in the user manual to remove references to Oracle generation and the Test Execution Engine, and explain how to use the new contract monitor. Also, update the disclaimer at the beginning of the section, as we now support structured types.

Jabe03 committed:
Updated docs with contract monitor section

Jabe03 committed:
Updated test generation page

Jabe03 committed:
contract monitor steps option allowed

Jabe03 committed:
Addressed some comments

Jabe03 committed:
Addressed more suggestions

Jabe03 commented:
I marked as draft because I am not done with a couple of the suggestions, and I would like to finalize the docs I wrote.

Jabe03 committed:
Meged contractMonitor module with interpreter

Jabe03 committed:
Removed the active modes list from lustrePath.t

Jabe03 committed:
Merge branch 'kind2-mc:develop' into josh-contract-monitor

Jabe03 committed:
removed now unused functions

Jabe03 committed:
Fixed error in displaying active modes

Jabe03 committed:
Merge branch 'kind2-mc:develop' into josh-contract-monitor

Jabe03 committed:
Updated docs for contract monitor and test generation

Jabe03 commented:
I have updated the docs with some changes to the test generation section

  - Now says JSON output instead of CSV
  - Removed references to the oracle and TEE
 
And the interpreter section

  - Added contract monitor as a subsection of the interpreter

I thought it may be good to just have the contract monitor section be a subsection of the interpreter since they are so similar and, by including it in the interpreter, the context is shared and there is less repeating of information. 
I am thinking about adding an example to the contract monitor section, but I am not sure if it would be redundant. In its current form, the contract monitor section is very short so maybe it is a good idea to have an example.

daniel-larraz committed:
Update test generation doc

daniel-larraz commented on src/kind2Flow.ml:
```diff
@@ -706,6 +708,29 @@ let run in_sys =
+        ISys.trans_sys_of_analysis
+          ~preserve_sig:true ~slice_nodes:`Off in_sys param
+      in
+      let sliced_inp = ISys.slice_to_abstraction in_sys param sys in 
```
First, you generate a transition system without slicing. Then, you compute a sliced version of the input system. Could you please clarify why this combination is needed?

Jabe03 replied:
This was an artifact of when we were trying to get this feature to work by slicing. I have removed the unnecessary input system computation.

daniel-larraz reviewed (changes requested):
Can you please update the schema files as well to include the new assumption, guarantee, and mode traces, and remove the `activeModes` section?

Jabe03 committed:
Remove unnecessary slicing

Jabe03 committed:
Empty contract traces excluded from output

Jabe03 committed:
Empty contract traces excluded from output (XML)

Jabe03 committed:
Schema update (JSON)

Jabe03 committed:
Update interpreter doc

Jabe03 committed:
Fix XML contract monitor type error

Jabe03 committed:
Update XML schema for contract monitor

Jabe03 committed:
Add class to stream types in contract traces in XML

Jabe03 committed:
Remove reference to active modes in JSON schema
